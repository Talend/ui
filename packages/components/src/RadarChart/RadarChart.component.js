import React from 'react';
import {
	RadarChart as RechartsRadarChart,
	PolarGrid,
	PolarRadiusAxis,
	Dot as RechartsDot,
	Radar,
	PolarAngleAxis,
} from 'recharts';
import PropTypes from 'prop-types';
import radarChartCssModule from './RadarChart.scss';
import { getTheme } from '../theme';

const theme = getTheme(radarChartCssModule);

/**
 * This function renders a Recharts radar chart for a given set of data
 * @param {Array} children required child elements
 * @param {string} className style classes for the component
 * @param {number} cx the horizontal start position of the chart
 * @param {number} cy the vertical start position of the chart
 * @param {Array} data the data to be rendered in the chart
 * @param {Array} domain the minimum and maximum chart values
 * @param {number} height the height of the chart container
 * @param {number} innerRadius the starting radius of the chart
 * @param {number} outerRadius the ending radius of the chart
 * @param {boolean} tick option to show a value scale on the chart
 * @param {boolean} tickLine option to draw a line from the chart to the tick/label
 * @param {number} width the width of the chart container
 */
export function RadarChart({
	children,
	className,
	cx,
	cy,
	data,
	domain,
	height,
	innerRadius,
	outerRadius,
	tick,
	tickLine,
	width,
}) {
	return (
		<RechartsRadarChart
			className={theme('tc-radar-chart', className)}
			cx={cx}
			cy={cy}
			outerRadius={outerRadius}
			innerRadius={innerRadius}
			width={width}
			height={height}
			data={data}
		>
			<PolarGrid />
			<PolarRadiusAxis domain={domain} tick={tick} axisLine={tickLine} />
			{children}
		</RechartsRadarChart>
	);
}
RadarChart.displayName = 'RadarChart';
RadarChart.propTypes = {
	children: PropTypes.node.isRequired,
	className: PropTypes.string,
	cx: PropTypes.number,
	cy: PropTypes.number,
	data: PropTypes.array.isRequired,
	domain: PropTypes.array,
	height: PropTypes.number,
	innerRadius: PropTypes.number,
	outerRadius: PropTypes.number,
	tick: PropTypes.bool,
	tickLine: PropTypes.bool,
	width: PropTypes.number,
};

RadarChart.defaultProps = {
	cx: 200,
	cy: 120,
	height: 240,
	innerRadius: 0,
	outerRadius: 100,
	tick: false,
	tickLine: false,
	width: 400,
};

/**
 * This function provides a custom clickable axis label with a data link to the index of the chart data
 * @param {Object} props the current props of the PolarAngleAxis
 */
// TODO 6.0: remove this export
export function LabelWithClick(props) {
	const { activeAxis, className, index, payload, y, ...rest } = props;
	let selectedClass = '';

	if (activeAxis === index) {
		selectedClass = theme('tc-radar-chart-label--selected', className);
	}

	return (
		<text {...rest} y={y + 3} data-axis-index={index} role="button" className={selectedClass}>
			{payload.value}
		</text>
	);
}
LabelWithClick.displayName = 'LabelWithClick';
LabelWithClick.propTypes = {
	activeAxis: PropTypes.number,
	className: PropTypes.string,
	index: PropTypes.number,
	payload: PropTypes.object,
	y: PropTypes.number,
};

/**
 * This function provides a custom dot with a data link to the index of the chart data
 * @param {Object} props the current props of the Radar
 */
// TODO 6.0: remove this export
export function Dot(props) {
	const { activeAxis, index } = props;
	let newR = 4;

	if (activeAxis === index) {
		newR = 4;
	} else {
		newR = 2;
	}
	return <RechartsDot {...props} fillOpacity={1} r={newR} />;
}
Dot.displayName = 'Dot';
Dot.propTypes = {
	activeAxis: PropTypes.number,
	index: PropTypes.number,
};

// TODO 6.0: remove those exports
export { Radar, PolarAngleAxis };

RadarChart.LabelWithClick = LabelWithClick;
RadarChart.Dot = Dot;
RadarChart.Radar = Radar;
RadarChart.PolarAngleAxis = PolarAngleAxis;
