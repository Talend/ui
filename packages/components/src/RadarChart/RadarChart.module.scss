@use '~@talend/bootstrap-theme/src/theme/guidelines' as *;
@use '~@talend/design-tokens/lib/tokens';

.tc-radar-chart {
	:global {
		.recharts-polar-grid-concentric-polygon:not(:last-child) {
			display: none;
		}

		.recharts-polar-angle-axis-ticks {
			transform: scale(1.06);
			transform-origin: center;
		}
	}
}

.tc-radar-chart-label--selected {
	fill: tokens.$coral-color-charts-default-strong;
	font-weight: 600;
}
