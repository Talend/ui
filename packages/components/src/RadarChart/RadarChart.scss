.tc-radar-chart {
	:global {
		.recharts-polar-grid-concentric-polygon:not(:last-child) {
			display: none;
		}
		.recharts-polar-angle-axis-ticks {
			transform: scale(1.06);
			transform-origin: center;
		}
	}
}

.tc-radar-chart-label--selected {
	fill: $regal-blue;
	font-weight: 600;
}
