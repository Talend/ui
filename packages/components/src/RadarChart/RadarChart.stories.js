import React from 'react';
import { storiesOf } from '@storybook/react';
import RadarChart from '.';

const ExampleDataSingle = [
	{ axis: 'Validity', A: 4 },
	{ axis: 'Popularity', A: 3 },
	{ axis: 'Completeness', A: 2 },
	{ axis: 'Discoverability', A: 1 },
	{ axis: 'other', A: 1 },
];

const ExampleDataMultiple = [
	{ axis: 'Validity', A: 4, B: 2 },
	{ axis: 'Popularity', A: 3, B: 5 },
	{ axis: 'Completeness', A: 2, B: 1 },
	{ axis: 'Discoverability', A: 1, B: 3 },
	{ axis: 'other', A: 1, B: 3 },
];

const ExampleDataClickable = [
	{ axis: 'Validity', A: 4 },
	{ axis: 'Popularity', A: 3 },
	{ axis: 'Completeness', A: 2 },
	{ axis: 'Discoverability', A: 1 },
];

const chartDomain = [0, 5];
const activeAxis = 2;

/**
 * Click functions provided by the host app
 */
const onClick = function () {
	document.getElementById('clickMessage').innerHTML =
		'Good job! You clicked ' + event.target.innerHTML + '.';
};

const stories = storiesOf('Data/Dataviz/RadarChart', module);
stories
	.addDecorator(story => <div className="col-lg-12">{story()}</div>)
	.add('Radar Chart Single', () => (
		<div>
			<h2>Single Object</h2>
			<p>A radar chart showing axis scores for a single object</p>
			<RadarChart data={ExampleDataSingle} domain={chartDomain}>
				<RadarChart.PolarAngleAxis dataKey="axis" />
				<RadarChart.Radar
					name="Trust score"
					dataKey="A"
					dot={true}
					stroke="#19426c"
					fill="#19426c"
					fillOpacity={0.1}
				/>
			</RadarChart>
		</div>
	))
	.add('Radar Chart Multiple', () => (
		<div>
			<h2>Multiple Objects</h2>
			<p>A radar chart showing axis scores for more than one object</p>
			<RadarChart data={ExampleDataMultiple} domain={chartDomain}>
				<RadarChart.PolarAngleAxis dataKey="axis" />
				<RadarChart.Radar
					name="Trust score 2019"
					dataKey="A"
					dot={false}
					stroke="#19426c"
					fill="#19426c"
					fillOpacity={0.1}
				/>
				<RadarChart.Radar
					name="Trust score 2020"
					dataKey="B"
					dot={false}
					stroke="#EA8330"
					fill="#EA8330"
					fillOpacity={0.1}
				/>
			</RadarChart>
		</div>
	))
	.add('Radar Chart Clickable', () => (
		<div>
			<h2>Clickable labels</h2>
			<p>A radar chart with clickable axis labels</p>
			<RadarChart data={ExampleDataClickable} domain={chartDomain}>
				<RadarChart.PolarAngleAxis
					dataKey="axis"
					tick={<RadarChart.LabelWithClick activeAxis={2} />}
					onClick={onClick}
				/>
				<RadarChart.Radar
					name="Trust score"
					dataKey="A"
					dot={false}
					stroke="#19426c"
					fill="#19426c"
					fillOpacity={0.1}
				/>
			</RadarChart>

			<div>
				<p id="clickMessage">You haven't clicked anything yet.</p>
			</div>
		</div>
	))
	.add('Radar Chart Custom Dot', () => (
		<div>
			<h2>Custom dot element</h2>
			<p>A radar chart with custom dots</p>
			<RadarChart data={ExampleDataClickable} domain={chartDomain}>
				<RadarChart.PolarAngleAxis dataKey="axis" />
				<RadarChart.Radar
					dataKey="A"
					dot={<RadarChart.Dot activeAxis={2} />}
					fill="#19426c"
					fillOpacity={0.1}
					isAnimationActive={false}
					name="Trust score"
					stroke="#19426c"
				/>
			</RadarChart>

			<div>
				<p id="clickMessage"></p>
			</div>
		</div>
	));
