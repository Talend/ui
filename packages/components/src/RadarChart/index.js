import { RadarChart } from './RadarChart.component';

export default RadarChart;
