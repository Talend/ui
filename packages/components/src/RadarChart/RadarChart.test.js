import React from 'react';
import { mount } from 'enzyme';
import RadarChart from '.';

describe('RadarChart', () => {
	it('should render a RadarChart', () => {
		const props = {
			data: [
				{ axis: 'Validity', A: 4 },
				{ axis: 'Social curation', A: 3 },
				{ axis: 'Completeness', A: 2 },
				{ axis: 'Discoverability', A: 1 },
				{ axis: 'other', A: 1 },
			],
			domain: [0, 5],
			dataKey: 'axis',
		};

		// when
		const wrapper = mount(
			<RadarChart data={props.data} domain={props.domain}>
				<RadarChart.PolarAngleAxis dataKey={props.dataKey} />
				<RadarChart.Radar
					name="Trust score"
					dataKey="A"
					dot
					stroke="#19426c"
					fill="#19426c"
					fillOpacity={0.1}
				/>
			</RadarChart>,
		);

		expect(wrapper.getElement()).toMatchSnapshot();
	});
	it('should render a chart with clickable labels', () => {
		const props = {
			data: [
				{ axis: 'Validity', A: 4 },
				{ axis: 'Social curation', A: 3 },
				{ axis: 'Completeness', A: 2 },
				{ axis: 'Discoverability', A: 1 },
				{ axis: 'other', A: 1 },
			],
			domain: [0, 5],
			dataKey: 'axis',
			activeAxis: 2,
			clickMock: jest.fn(),
		};

		// when
		const wrapper = mount(
			<RadarChart data={props.data} domain={props.domain}>
				<RadarChart.PolarAngleAxis
					dataKey="axis"
					tick={<RadarChart.LabelWithClick activeAxis={props.activeAxis} />}
					onClick={props.clickMock}
				/>
				<RadarChart.Radar
					name="Trust score"
					dataKey="A"
					dot={false}
					stroke="#19426c"
					fill="#19426c"
					fillOpacity={0.1}
				/>
			</RadarChart>,
		);

		// when
		wrapper.find('.recharts-polar-angle-axis-tick').at(0).simulate('click');

		// then
		expect(props.clickMock).toHaveBeenCalled();
	});
	it('should render a chart with custom dots', () => {
		const props = {
			data: [
				{ axis: 'Validity', A: 4 },
				{ axis: 'Social curation', A: 3 },
				{ axis: 'Completeness', A: 2 },
				{ axis: 'Discoverability', A: 1 },
				{ axis: 'other', A: 1 },
			],
			domain: [0, 5],
			dataKey: 'axis',
			activeAxis: 2,
		};

		// when
		const wrapper = mount(
			<RadarChart data={props.data} domain={props.domain}>
				<RadarChart.PolarAngleAxis dataKey={props.dataKey} />
				<RadarChart.Radar
					dataKey="A"
					dot={<RadarChart.Dot activeAxis={props.activeAxis} />}
					fill="#19426c"
					fillOpacity={0.1}
					isAnimationActive={false}
					name="Trust score"
					stroke="#19426c"
				/>
			</RadarChart>,
		);

		expect(wrapper.getElement()).toMatchSnapshot();
	});
});
