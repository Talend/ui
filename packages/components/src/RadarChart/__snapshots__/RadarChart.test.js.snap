// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadarChart should render a RadarChart 1`] = `
<RadarChart
  cx={210}
  cy={140}
  data={
    Array [
      Object {
        "A": 4,
        "axis": "Validity",
      },
      Object {
        "A": 3,
        "axis": "Social curation",
      },
      Object {
        "A": 2,
        "axis": "Completeness",
      },
      Object {
        "A": 1,
        "axis": "Discoverability",
      },
      Object {
        "A": 1,
        "axis": "other",
      },
    ]
  }
  domain={
    Array [
      0,
      5,
    ]
  }
  height={250}
  innerRadius={0}
  outerRadius={100}
  tick={false}
  tickLine={false}
  width={420}
>
  <PolarAngleAxis
    allowDuplicatedCategory={true}
    angleAxisId={0}
    axisLine={true}
    cx={0}
    cy={0}
    dataKey="axis"
    domain={
      Array [
        0,
        "auto",
      ]
    }
    hide={false}
    orientation="outer"
    scale="auto"
    tick={true}
    tickLine={true}
    type="category"
  />
  <Radar
    activeDot={true}
    angleAxisId={0}
    animationBegin={0}
    animationDuration={1500}
    animationEasing="ease"
    dataKey="A"
    dot={true}
    fill="#19426c"
    fillOpacity={0.1}
    hide={false}
    isAnimationActive={true}
    legendType="rect"
    name="Trust score"
    radiusAxisId={0}
    stroke="#19426c"
  />
</RadarChart>
`;

exports[`RadarChart should render a chart with custom dots 1`] = `
<RadarChart
  cx={210}
  cy={140}
  data={
    Array [
      Object {
        "A": 4,
        "axis": "Validity",
      },
      Object {
        "A": 3,
        "axis": "Social curation",
      },
      Object {
        "A": 2,
        "axis": "Completeness",
      },
      Object {
        "A": 1,
        "axis": "Discoverability",
      },
      Object {
        "A": 1,
        "axis": "other",
      },
    ]
  }
  domain={
    Array [
      0,
      5,
    ]
  }
  height={250}
  innerRadius={0}
  outerRadius={100}
  tick={false}
  tickLine={false}
  width={420}
>
  <PolarAngleAxis
    allowDuplicatedCategory={true}
    angleAxisId={0}
    axisLine={true}
    cx={0}
    cy={0}
    dataKey="axis"
    domain={
      Array [
        0,
        "auto",
      ]
    }
    hide={false}
    orientation="outer"
    scale="auto"
    tick={true}
    tickLine={true}
    type="category"
  />
  <Radar
    activeDot={true}
    angleAxisId={0}
    animationBegin={0}
    animationDuration={1500}
    animationEasing="ease"
    dataKey="A"
    dot={
      <Dot
        activeAxis={2}
      />
    }
    fill="#19426c"
    fillOpacity={0.1}
    hide={false}
    isAnimationActive={false}
    legendType="rect"
    name="Trust score"
    radiusAxisId={0}
    stroke="#19426c"
  />
</RadarChart>
`;
